### PR TITLE
feat: ``function/id/run`` route

### PR DESCRIPTION
**Pull request by WoodenMaiden**

### CODE REVIEW

File: routes/runFunction.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	// put these two into gofuncs
+	if tx.Error != nil {
+		// todo
+		// request a function to the scheduler here
+		// fn = something
+	}
+
+	if fn.Status == "Pending" {
+		// scheduler.WaitForReadiness(fn.Id)
+		// Maybe add a pubsub in the PG that triggers when a function is ready aka when the status is updated, https://www.postgresql.org/docs/current/sql-notify.html
```
Comment (WoodenMaiden) [inline routes/runFunction.go:38]: @nponsard @alexis-langlet what do you think ?

Comment (alexis-langlet) [reply routes/runFunction.go:38]: well, for now I'm not sure implementing a pubsub is a great idea.
Let's begin with simple stuff and send a request to check readiness every X ms at first
We will improve that if needed and if time allows it

Comment (nponsard) [reply routes/runFunction.go:38]: I second this

**Review by nponsard — DISMISSED** [review]

**Comment by WoodenMaiden** [comment]
depends on #7

File: database/functionState.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+type FunctionState struct {
+	gorm.Model
+	ID      uuid.UUID `json:"id" ,gorm:"primarykey;type:uuid;default:gen_random_uuid()"`
+	Code    string    `json:"code"`
+	Status  string    `json:"status"`
+	Address string    `json:"address"`
+	Port    uint16    `json:"port"`
+	OwnerID int       `json:"owner_id"`
+	Owner   User	  `json:"-"`
+}
```
[the remark this answers is not in the document]
Comment (WoodenMaiden) [reply database/functionState.go]: fixed in 4a86caa292a1d9a82bdcdaf0500745f24dee62bb feel free to check it out!

File: routes/function/controller.go [hunk 3]
```diff
@@ -164,3 +169,93 @@ type BuilderRequest struct {
...
+		}
+
+		// if even after 5 attempts the function is not ready, we return an error
+		if fnState.Status != "Ready" {
+			ctx.AbortWithStatusJSON(500, gin.H{"error": "Function is not ready"})
+			return
+		}
+	}
+
+	// TODO: forward the body to fn.Address:fn.Port
```
Comment (nponsard) [inline routes/function/controller.go:258]: I think you forgot to implement this

Comment (WoodenMaiden) [reply routes/function/controller.go:258]: oops

File: routes/function/controller.go [hunk 4]
```diff
@@ -167,3 +171,121 @@ type BuilderRequest struct {
...
+	if err != nil {
+		log.Println(fmt.Sprint("Could not update state of VM", fnState.ID ,": ", err.Error()))
+		ctx.AbortWithStatusJSON(500, gin.H{"error": "Cannot update function's status"})
+		return
+	}
+	
+	_, err = http.Post(
+		fmt.Sprint("http://", string(fnState.Address), ":", fnState.Port, "/execute"),
+		"application/json",
+		bytes.NewReader(fnBody),
```
Comment (nponsard) [inline routes/function/controller.go:270]: You could pass `ctx.Request.Body` instead, avoids loading the body in memory

**Review by nponsard — COMMENTED** [review]
Otherwise looks good

**Pull request merged**